core/src/main/scala/kafka/log/Log.scala

```diff
@@ -717,16 +746,20 @@ class Log(@volatile var dir: File,
     deleteOldSegments(shouldDelete)
   }
 
+  /*
+   * The log's next offset may be smaller than log start offset when the log is purged upon PurgeRequest
+   * We don't need to call this method periodically using scheduler since it is called right after broker receives PurgeRequest
+   */
+  private def deleteLogStartOffsetBreachedSegments() : Int = {
+    // keep active segment to avoid frequent log rolling due to user's PurgeRequest
+    deleteOldSegments(segment => segment.nextOffset <= logStartOffset && segment.baseOffset < activeSegment.baseOffset)
```

Comment:
It seems we should be careful about when to delete the log segments. In the current code, when the leader deleted an old log segment. The low watermark will not increase immediately because the followers has not updated the logStartOffset yet. However, at this point if user tries to fetch from LW, they will receive an offset out of range exception, which is not expected. 

One solution would be letting the segment deletion to be based on the low watermark instead of logStartOffset. So we need to delay the log deletion until low watermark is updated. Similarly the valid range of fetch request should be based on LW instead of logStartOffset.

Nit: There is an existing typo in line 729, which missed a "t" in "deleteRetentionMsBreachedSegments()".